### PR TITLE
Fix mask propagation in the presence of where

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125574
* #126783
* #114471

Before, when calling ops.where, masks were not properly propagated. We
now restrict the optimisation to `ops.masked`, which I think it was what
the original code intended to do.

I'm not 100% sure that even in the masked case this code is not
introducing some bugs, but this is a strict improvement over the
previous state.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -471,10 +471,12 @@ def __init__(self, name, bounds: ValueRanges[Any]):
...
     def update_on_args(self, name, args, kwargs):
         # When making a variable that is going to be used in indirect indexing
-        # if a where clause is used it should mean that the result is always a
+        # if we are within a masked block, it means that the result is always a
         # valid index, so you shouldn't include any of the dependent variables
         # in the resulting load mask
-        if name == "where":
+        if (
+            fx_node := getattr(V.interpreter, "current_node", None)
+        ) and fx_node.target.startswith("masked"):
```
Comment (peterbell10) [inline torch/_inductor/codegen/triton.py:479]: How does this interact with the masked sub-block? Is current_node only masked for the output of `ops.masked`?

Also pre-existing, but why do we exclude `xmask`? Shouldn't this only be filtering out `tmp{n}` masks?

Comment (lezcano) [reply torch/_inductor/codegen/triton.py]: It did interact alright, but now the implementation is even cleaner without using all those global objects.

**Comment by lezcano** [comment]
An even simpler implementation.
It'd be great to get rid of `update_on_args`, as it's quite easy to use incorrectly (you have to remember to call it on every `ops` call you make during codegen). We can certainly get rid of it on the triton backend by lifting the mask propagation logic to the ops handler. The C++ backend could also get rid of it once they transition their dtype propagation to use something like https://github.com/pytorch/pytorch/pull/123100 (cc @jgong5)

**Review by peterbell10 — COMMENTED** [review]
Makes sense, can you add a test though?

**Comment by lezcano** [comment]
It's quite tricky to hit this bug. I've been trying for some time without luck. I think a potential case would be if you use `torch.where` inside an `ops.masked` and then you use this within an indirect assert. Do you have a simpler example that would exercise this code?

**Comment by peterbell10** [comment]
I think this is a minimal reproducer. 
```python
import torch

def fn(a):
    idx = torch.arange(a.size(0), device=a.device)
    padded_idx = torch.constant_pad_nd(idx, (1050, 0))
    padded_idx = torch.where(padded_idx >= 0, padded_idx, padded_idx)
    return a[padded_idx]

a = torch.randn(1024, device="cuda")
cfn = torch.compile(fn)
cfn(a)
```
It generates the assert:
```python
tl.device_assert((0 <= tmp9) & (tmp9 < 1024), "index out of bounds: 0 <= tmp9 < 1024")
```

which has dropped the `xmask`.

**Comment by lezcano** [comment]
trigger a merge when you approve if nothing looks wrong please

**Review by peterbell10 — APPROVED** [review]
